servers/src/main/java/tachyon/worker/block/BlockMasterSync.java

```diff
@@ -103,25 +103,54 @@ public void setWorkerId() throws IOException {
   /**
    * Registers with the Tachyon master. This should be called before the continuous heartbeat thread
    * begins. The workerId will be set after this method is successful.
+   *
+   * @throws IOException when current workerId cannot be found in master
    */
-  public void registerWithMaster() {
+  public void registerWithMaster() throws IOException {
     BlockStoreMeta storeMeta = mBlockDataManager.getStoreMeta();
     try {
       mMasterClient.register(mWorkerId, storeMeta.getCapacityBytesOnTiers(),
           storeMeta.getUsedBytesOnTiers(), storeMeta.getBlockList());
     } catch (IOException ioe) {
-      throw new RuntimeException("Failed to register with master.", ioe);
+      LOG.error("Failed to register to master", ioe);
+      throw ioe;
     }
   }
 
   /**
+   * Try to get new id from master and then register again, when any of the two actions fails,
+   * connection to master will be reset, and the re-registration will be tried again after 1 second.
+   * When heartbeat timeout, a {@link RuntimeException} will be thrown.
+   *
+   * @param lastHeartbeatMs start time to compute elapsed time
+   */
+  public void reregisterWithMaster(long lastHeartbeatMs) {
```

Comment:
I'm not sure if we have to actively retry to reregister with the master. The sync thread already has a loop, which continually heartbeats (or registers again if commanded), so that will take care of the retries.

I also think that we won't need a new method reregisterWithMaster(). This is because when the heartbeat rpc is attempted, it will send back a command to re-register if the workerId doesn't exist. That command will call setWorkerId() and then register() again. If the register didn't work, it will just get a re-register command to do it again.